wp-content/plugins/wporg-learn/inc/capabilities.php

```diff
@@ -0,0 +1,237 @@
+<?php
+
+namespace WPOrg_Learn\Capabilities;
+
+defined( 'WPINC' ) || die();
+
+/**
+ * Actions and filters.
+ */
+add_filter( 'user_has_cap', __NAMESPACE__ . '\set_post_type_caps' );
+add_filter( 'user_has_cap', __NAMESPACE__ . '\set_caps_for_internal_notes' );
+add_filter( 'map_meta_cap', __NAMESPACE__ . '\map_meta_caps', 20, 4 ); // Needs to fire after meta caps in wporg-internal-notes.
+add_action( 'init', __NAMESPACE__ . '\add_or_update_lesson_plan_editor_role' );
+add_action( 'init', __NAMESPACE__ . '\add_or_update_workshop_reviewer_role' );
+
+/**
+ * Assign custom post type caps to roles based on their equivalents for the 'post' post type.
+ *
+ * Example: If a user has the `edit_others_posts` cap (from Editor role), this will also give them
+ * the equivalent `edit_others_lesson_plans` and `edit_others_workshops` caps.
+ *
+ * @param bool[] $user_caps A list of primitive caps (keys) and whether user has them (boolean values).
+ *
+ * @return array
+ */
+function set_post_type_caps( $user_caps ) {
+	$capability_types = array(
+		array( 'lesson_plan', 'lesson_plans' ),
+		array( 'workshop', 'workshops' ),
+	);
+
+	foreach ( $capability_types as $capability_type ) {
+		// Generate the caps for a capability type.
+		$cap_args = array(
+			'capability_type' => $capability_type,
+			'capabilities'    => array(),
+			'map_meta_cap'    => true,
+		);
+		$cap_map = (array) get_post_type_capabilities( (object) $cap_args );
+
+		foreach ( $user_caps as $cap => $bool ) {
+			if ( $bool && isset( $cap_map[ $cap ] ) ) {
+				$user_caps[ $cap_map[ $cap ] ] = true;
+			}
+		}
+	}
+
+	return $user_caps;
+}
+
+/**
+ * Enable a cap for managing internal notes on workshop posts.
+ *
+ * The `promote_users` cap is shared by admin and workshop reviewer roles, which are the two roles
+ * that should also have access to internal notes.
+ *
+ * @param bool[] $user_caps
+ *
+ * @return mixed
+ */
+function set_caps_for_internal_notes( $user_caps ) {
+	if ( isset( $user_caps['promote_users'] ) && true === $user_caps['promote_users'] ) {
+		$user_caps['manage_workshop_internal_notes'] = true;
+	}
+
+	return $user_caps;
+}
+
+/**
+ * Map primitive caps to our custom caps.
+ *
+ * @param array  $required_caps
+ * @param string $current_cap
+ * @param int    $user_id
+ * @param mixed  $args
+ *
+ * @return mixed
+ */
+function map_meta_caps( $required_caps, $current_cap, $user_id, $args ) {
+	switch ( $current_cap ) {
+		case 'edit_any_learn_content':
+			$required_caps       = array();
+			$learn_content_types = array( 'lesson-plan', 'wporg_workshop', 'course', 'lesson' );
+
+			foreach ( $learn_content_types as $post_type ) {
+				$object = get_post_type_object( $post_type );
+				if ( user_can( $user_id, $object->cap->edit_posts ) ) {
+					$required_caps[] = $object->cap->edit_posts;
+					break 2;
```

Comment:
```suggestion
					return $required_caps;
```

In general I think single points of return are simpler, but in this case an explicit `return` here feels much simpler than `break 2`.

Reply:
Feels wrong to return early, but I added a comment to clarify what we're breaking out of in bf24d13